File: linda-core/src/main/java/it/unibo/coordination/linda/core/BulkTupleSpace.java
package it.unibo.coordination.linda.core;

import org.apache.commons.collections4.MultiSet;
import org.apache.commons.collections4.multiset.HashMultiSet;

import java.util.Collection;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;

import static it.unibo.coordination.utils.ListUtils.listOf;

public interface BulkTupleSpace<T extends Tuple, TT extends Template, K, V> extends TupleSpace<T, TT, K, V> {

    CompletableFuture<Collection<? extends Match<T, TT, K, V>>> readAll(TT template);

    default CompletableFuture<MultiSet<T>> readAllTuples(TT template) {
        return readAll(template).thenApplyAsync(matches ->
                    matches.stream().map(m -> m.getTuple().get())
                        .collect(Collectors.toCollection(HashMultiSet::new))
                );
    }

    CompletableFuture<Collection<? extends Match<T, TT, K, V>>> takeAll(TT template);

    default CompletableFuture<MultiSet<T>> takeAllTuples(TT template) {
        return takeAll(template).thenApplyAsync(matches ->
                matches.stream().map(m -> m.getTuple().get())
                        .collect(Collectors.toCollection(HashMultiSet::new))
        );
    }

    CompletableFuture<MultiSet<T>> writeAll(Collection<? extends T> tuples);

    default CompletableFuture<MultiSet<T>> writeAll(final T tuple1, final T... otherTuples) {
        return writeAll(listOf(tuple1, otherTuples));
    }
}

File: linda-core/src/main/kotlin/it/unibo/coordination/linda/core/BulkTupleSpace.kt
package it.unibo.coordination.linda.core

import it.unibo.coordination.Promise
import org.apache.commons.collections4.multiset.HashMultiSet

interface BulkTupleSpace<T : Tuple, TT : Template, K, V> : LindaTupleSpace<T, TT, K, V> {

    fun readAll(template: TT): Promise<Collection<Match<T, TT, K, V>>>

    @JvmDefault
    fun readAll(template: String): Promise<Collection<Match<T, TT, K, V>>> =
            readAll(template.toTemplate())

    @JvmDefault
    fun readAllTuples(template: TT): Promise<Collection<T>> {
        return readAll(template).thenApplyAsync { matches ->
            HashMultiSet(matches.map { it.tuple.get() })
        }
    }

    @JvmDefault
    fun readAllTuples(template: String): Promise<Collection<T>> =
            readAllTuples(template.toTemplate())

    fun takeAll(template: TT): Promise<Collection<Match<T, TT, K, V>>>

    @JvmDefault
    fun takeAll(template: String): Promise<Collection<Match<T, TT, K, V>>> =
            takeAll(template.toTemplate())

    @JvmDefault
    fun takeAllTuples(template: TT): Promise<Collection<T>> {
        return takeAll(template).thenApplyAsync { matches ->
            HashMultiSet(matches.map { it.tuple.get() })
        }
    }

    @JvmDefault
    fun takeAllTuples(template: String): Promise<Collection<T>> =
            takeAllTuples(template.toTemplate())

    fun writeAll(tuples: Collection<T>): Promise<Collection<T>>

    @JvmDefault
    fun writeAll(tuple1: String, vararg otherTuples: String): Promise<Collection<T>> =
            writeAll(listOf(tuple1, *otherTuples).map { it.toTuple() })

    @JvmDefault
    fun writeAll(tuple1: T, vararg otherTuples: T): Promise<Collection<T>> {
        return writeAll(listOf(tuple1, *otherTuples))
    }
}